cratedb_toolkit/io/mongodb/export.py

```diff
@@ -56,36 +65,157 @@ def timestamp_converter(value):
-        if len(value) == 1:
+    def apply_special_treatments(self, value):
+        """
+        Apply special treatments to value that can't be described otherwise up until now.
+        # Ignore certain items including anomalies that are not resolved, yet.
+
+        TODO: Needs an integration test feeding two records instead of just one.
+        """
+
+        if self.config is None or self.config.treatment is None:
+            return
+
+        # Optionally ignore lists of complex objects.
+        local_ignores = []
+        if self.config.treatment.ignore_complex_lists:
             for k, v in value.items():
-                if k.startswith("$"):
-                    return extract_value(v, k.lstrip("$"))
-        return {k.lstrip("$"): extract_value(v, parent_type) for (k, v) in value.items()}
-    if isinstance(value, list):
-        return [extract_value(v, parent_type) for v in value]
-    if parent_type:
-        converter = type_converter.get(parent_type)
-        if converter:
-            return converter(value)
-    return value
+                if isinstance(v, list) and v and isinstance(v[0], dict):
+                    # Skip ignoring special-encoded items.
+                    if v[0] and list(v[0].keys())[0].startswith("$"):
+                        continue
+                    local_ignores.append(k)
+
+        # Apply global and computed ignores.
+        for ignore_name in self.config.treatment.ignore_columns + local_ignores:
+            if ignore_name in value:
+                del value[ignore_name]
+
+        # Converge certain items to `list` even when defined differently.
+        for to_list_name in self.config.treatment.to_list:
+            if to_list_name in value and not isinstance(value[to_list_name], list):
+                value[to_list_name] = [value[to_list_name]]
+
+        # Converge certain items to `str` even when defined differently.
+        for name in self.config.treatment.to_string:
+            if name in value and not isinstance(value[name], str):
+                value[name] = str(value[name])
+
+        """
+        # Manual treatment.
+        # Some nested objects have been defined as strings, probably in previous schema versions.
+        if "users" in value:
+            for user_item in value["users"]:
+                if "user" in user_item and not isinstance(user_item["user"], dict):
+                    user_item["user"] = {"id": user_item["user"]}
+        """
+
+        if "createdBy" in value and not isinstance(value["createdBy"], dict):
+            value["createdBy"] = {"id": value["createdBy"]}
+
+        # Prune invalid date representations.
+        for key in ["start_date", "end_date"]:
+            if key in value:
+                if not isinstance(value[key], dict):
+                    del value[key]
+                elif "date" in value[key]:
+                    if isinstance(value[key]["date"], str):
+                        del value[key]
```

Comment:
While others are configurable per `--treatment=Path` option using a corresponding YAML file, those special treatments have not been generalized yet.

Reply:
Those rules now have been generalized further to be controlled using the treatment file.
```yaml
convert_dict:
  - name: createdBy
    wrapper_name: id
  - name: user
    wrapper_name: id
prune_invalid_date:
  - "start_date"
  - "end_date"
```

Reply:
The `--treatment` option has been dissolved again. Special treatments have been integrated into Zyp.

- https://github.com/crate/commons-codec/pull/46